src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,343 @@
+        Ok(S3MultiPartUploader {
+            client: crate::s3::new_client(sdk_config),
+            bucket,
+            path,
+            state: Initialized,
+        })
+    }
+}
+
+impl S3MultiPartUploader<Initialized> {
+    /// Starts the multi-part upload process by making a `create_multipart_upload` request to s3.
+    pub async fn start(self) -> Result<S3MultiPartUploader<InProgress>, S3MultiPartUploadError> {
+        let res = self
+            .client
+            .create_multipart_upload()
+            .bucket(&self.bucket)
+            .key(&self.path)
+            .send()
+            .await?;
+
+        let upload_id = res
+            .upload_id()
+            .ok_or(anyhow!(
+                "create_multipart_upload response missing upload id"
+            ))?
+            .to_string();
+
+        Ok(S3MultiPartUploader {
+            client: self.client,
+            bucket: self.bucket,
+            path: self.path,
+            state: InProgress {
+                upload_id,
+                current_part_number: 0,
+                etags: BTreeMap::new(),
+                total_bytes_uploaded: 0,
+            },
+        })
+    }
+}
+
+impl S3MultiPartUploader<InProgress> {
+    async fn upload_part_internal(
+        &mut self,
+        data: Bytes,
+    ) -> Result<MultiPartProgress, S3MultiPartUploadError> {
+        let num_of_bytes: u64 = data.len().try_into().expect("usize to u64");
+
+        let next_part_number = self.state.current_part_number + 1;
+        if next_part_number > MAX_PART_NUM {
+            return Err(S3MultiPartUploadError::ExceedsMaxPartNumber);
+        }
+
+        let res = self
+            .client
+            .upload_part()
+            .bucket(&self.bucket)
+            .key(&self.path)
+            .upload_id(self.state.upload_id.clone())
+            .part_number(next_part_number)
+            .body(ByteStream::from(data))
+            .send()
+            .await?;
+
+        self.state.etags.insert(
+            next_part_number,
+            res.e_tag
+                .ok_or(anyhow!("upload_part response missing etag"))?,
+        );
+        self.state.current_part_number = next_part_number;
+        self.state.total_bytes_uploaded += num_of_bytes;
+        Ok(MultiPartProgress {
+            num_of_parts: self
+                .state
+                .current_part_number
+                .try_into()
+                .expect("i32 to u32"),
+            total_bytes_uploaded: self.state.total_bytes_uploaded,
+        })
+    }
+
+    /// Uploads a part for an ongoing multi-part upload.
+    /// Can only be called after calling `S3MultiPartUploader::start`
+    /// and can be called as many times required to upload each individual parts.
+    /// The total number of parts cannot exceed `MAX_PART_NUM` and
+    /// each part size should at least be `MIN_PART_SIZE` in size.
+    /// Returns the number of parts and number of bytes uploaded till now.
```

Comment:
Recommend abstracting over this complexity. I'd recommend that `upload_part` accept chunks of any size, and accumulate them into a local buffer. Then, whenever you've exceeded the minimum part size (and honestly we should probably pick a larger number than 5MB), you flush out the accumulated chunks as a single part to AWS. 

Then, `finish` doesn't need to accept an `Option<Bytes>`. It just needs to call `flush` on whatever buffer is accumulated, if any.

Reply:
I see what you mean. Updating